spec-main/api-browser-window-spec.ts

```diff
@@ -2590,65 +2591,112 @@ describe('BrowserWindow module', () => {
       w.loadFile(path.join(__dirname, 'fixtures', 'api', 'close-beforeunload-undefined.html'));
     });
     it('returning false would prevent close', async () => {
-      w.loadFile(path.join(__dirname, 'fixtures', 'api', 'close-beforeunload-false.html'));
+      await w.loadFile(path.join(__dirname, 'fixtures', 'api', 'close-beforeunload-false.html'));
+      w.webContents.executeJavaScript('run()', true);
       const [e] = await emittedOnce(ipcMain, 'onbeforeunload');
       e.returnValue = null;
     });
     it('returning empty string would prevent close', (done) => {
       ipcMain.once('onbeforeunload', (e) => { e.returnValue = null; done(); });
       w.loadFile(path.join(__dirname, 'fixtures', 'api', 'close-beforeunload-empty-string.html'));
     });
-    it('emits for each close attempt', (done) => {
-      let beforeUnloadCount = 0;
-      ipcMain.on('onbeforeunload', (e) => {
+
+    it('emits for each close attempt', async () => {
+      await w.loadFile(path.join(__dirname, 'fixtures', 'api', 'beforeunload-false-prevent3.html'));
+
+      const destroyListener = () => { expect.fail('Close was not prevented'); };
+      w.webContents.once('destroyed', destroyListener);
+
+      await w.webContents.executeJavaScript('preventNextBeforeUnload()', true);
+      {
+        const p = emittedOnce(ipcMain, 'onbeforeunload');
+        w.close();
+        const [e] = await p;
         e.returnValue = null;
-        beforeUnloadCount += 1;
-        if (beforeUnloadCount < 3) {
-          w.close();
-        } else if (beforeUnloadCount === 3) {
-          done();
-        }
-      });
-      w.webContents.once('did-finish-load', () => { w.webContents.executeJavaScript('window.close()', true); });
-      w.loadFile(path.join(__dirname, 'fixtures', 'api', 'beforeunload-false-prevent3.html'));
+      }
+
+      await w.webContents.executeJavaScript('preventNextBeforeUnload()', true);
+
+      // Hi future test refactorer! I don't know what event this timeout allows
+      // to occur, but without it, this test becomes flaky at this point and
+      // sometimes the window gets closed even though a `beforeunload` handler
+      // has been installed. I looked for events being emitted by the
+      // `webContents` during this timeout period and found nothing, so it
+      // might be some sort of internal timeout being applied by the content/
+      // layer, or blink?
```

Comment:
That would be https://source.chromium.org/chromium/chromium/src/+/master:content/browser/frame_host/render_frame_host_impl.cc;l=791-796 :)

Reply:
Looks like that's set to 500ms https://source.chromium.org/chromium/chromium/src/+/master:content/browser/renderer_host/render_view_host_impl.cc;l=143;drc=da6ef73266f6110abc5e788457f0e8f9257dd5b6;bpv=1;bpt=1

that explains why it didn't work when i made this `setTimeout(resolve, 0)`!